### PR TITLE
Integrate Graph Architecture with rest of codebase

### PR DESCRIPTION
**Pull request by wochinge**
**Proposed changes**:
- fix https://github.com/RasaHQ/rasa/issues/9277
- fix https://github.com/RasaHQ/rasa/issues/9604
- integrate the Graph Architecture with the rest of the codebase

**Open Issues**
- [x] https://github.com/RasaHQ/rasa/issues/9765
- [x] https://github.com/RasaHQ/rasa/issues/9713
- [x] https://github.com/RasaHQ/rasa/issues/9766
- [x] https://github.com/RasaHQ/rasa/issues/9767

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/graph_components/validators/default_recipe_validator.py [hunk 1]
```diff
@@ -48,7 +49,7 @@
 import rasa.shared.utils.io
 
 
-# TODO: replace these once the Recipe is merged
+# TODO: Can we replace this with the registered types from the regitry?
```
Comment (wochinge) [inline rasa/graph_components/validators/default_recipe_validator.py:52]: Unfortunately this doesn't work as expected:
* `EntitySynonymMapper` is a `EntityExtractor` in the recipe but not part of this
* `DIETClassifier` is `IntentClassifier` in the recipe but a `EntityExtractor` here

**Comment by wochinge** [comment]
[Some model regression tests were failing over night due to poetry issues](https://github.com/RasaHQ/rasa/runs/3885042429?check_suite_focus=true). As we have also seen no changes in previous runs and the entity testing works now, I just give it one final run on the Sara dataset before merging @joejuzl @ka-bu

**Comment by koernerfelicia** [comment]
Run when compared to TF 2.6 (note 11% drop for Rules+Memo for Sara is not explained yet, ignore this)

Dataset: `Carbon Bot`, Dataset repository branch: `architecture-revamp/modified-spacy-configs`, commit: `a88501268464439d1d058a8b77f16a1c752230bd`
| Configuration | Intent Classification Micro F1 | Entity Recognition Micro F1 | Response Selection Micro F1 |
|---------------|-----------------|-----------------|-------------------|
| `BERT + DIET(bow) + ResponseSelector(bow)`<br> test: `1m28s`, train: `4m12s`, total: `5m40s`|0.7942 (0.00)|0.4235 (-0.33)|0.5382 (0.00)|
| `BERT + DIET(seq) + ResponseSelector(t2t)`<br> test: `1m41s`, train: `4m19s`, total: `5m59s`|0.8078 (0.01)|0.7787 (0.00)|0.5430 (0.01)|
| `Sparse + BERT + DIET(bow) + ResponseSelector(bow)`<br> test: `1m31s`, train: `4m54s`, total: `6m24s`|0.7689 (-0.03)|0.4235 (-0.33)|0.5497 (-0.01)|
| `Sparse + BERT + DIET(seq) + ResponseSelector(t2t)`<br> test: `1m48s`, train: `5m8s`, total: `6m55s`|0.7806 (0.00)|0.7880 (-0.01)|0.5497 (-0.03)|
| `Sparse + DIET(bow) + ResponseSelector(bow)`<br> test: `44s`, train: `2m50s`, total: `3m34s`|0.7398 (-0.01)|0.4235 (-0.33)|0.5249 (-0.01)|
| `Sparse + DIET(seq) + ResponseSelector(t2t)`<br> test: `1m3s`, train: `4m9s`, total: `5m11s`|0.7398 (0.00)|0.7022 (0.02)|0.5497 (0.01)|


Dataset: `Hermit`, Dataset repository branch: `architecture-revamp/modified-spacy-configs`, commit: `a88501268464439d1d058a8b77f16a1c752230bd`
| Configuration | Intent Classification Micro F1 | Entity Recognition Micro F1 | Response Selection Micro F1 |
|---------------|-----------------|-----------------|-------------------|
| `BERT + DIET(bow) + ResponseSelector(bow)`<br> test: `2m30s`, train: `21m21s`, total: `23m51s`|0.8978 (-0.00)|0.7504 (0.00)|`no data`|
| `BERT + DIET(seq) + ResponseSelector(t2t)`<br> test: `2m46s`, train: `15m1s`, total: `17m47s`|0.8913 (0.00)|0.8049 (0.00)|`no data`|
| `Sparse + BERT + DIET(bow) + ResponseSelector(bow)`<br> test: `2m40s`, train: `26m49s`, total: `29m28s`|0.8699 (-0.00)|0.7504 (0.00)|`no data`|
| `Sparse + BERT + DIET(seq) + ResponseSelector(t2t)`<br> test: `2m56s`, train: `19m0s`, total: `21m56s`|0.8727 (0.00)|0.7994 (0.00)|`no data`|
| `Sparse + DIET(bow) + ResponseSelector(bow)`<br> test: `1m6s`, train: `21m14s`, total: `22m19s`|0.8336 (0.00)|0.7504 (0.00)|`no data`|
| `Sparse + DIET(seq) + ResponseSelector(t2t)`<br> test: `1m21s`, train: `14m38s`, total: `15m58s`|0.8355 (0.00)|0.7561 (0.00)|`no data`|


Dataset: `Private 1`, Dataset repository branch: `architecture-revamp/modified-spacy-configs`, commit: `a88501268464439d1d058a8b77f16a1c752230bd`
| Configuration | Intent Classification Micro F1 | Entity Recognition Micro F1 | Response Selection Micro F1 |
|---------------|-----------------|-----------------|-------------------|
| `BERT + DIET(bow) + ResponseSelector(bow)`<br> test: `1m49s`, train: `4m1s`, total: `5m49s`|0.9096 (0.00)|0.9612 (0.00)|`no data`|
| `BERT + DIET(seq) + ResponseSelector(t2t)`<br> test: `2m10s`, train: `3m46s`, total: `5m56s`|0.9148 (0.00)|0.9717 (0.00)|`no data`|
| `Spacy + DIET(bow) + ResponseSelector(bow)`<br> test: `34s`, train: `3m2s`, total: `3m35s`|0.8420 (0.00)|0.9574 (0.00)|`no data`|
| `Spacy + DIET(seq) + ResponseSelector(t2t)`<br> test: `53s`, train: `3m25s`, total: `4m17s`|0.8534 (-0.00)|0.9444 (0.00)|`no data`|
| `Sparse + DIET(bow) + ResponseSelector(bow)`<br> test: `29s`, train: `3m37s`, total: `4m6s`|0.8929 (0.00)|0.9612 (0.00)|`no data`|
| `Sparse + DIET(seq) + ResponseSelector(t2t)`<br> test: `48s`, train: `3m32s`, total: `4m20s`|0.9085 (0.01)|0.9735 (0.00)|`no data`|
| `Sparse + Spacy + DIET(bow) + ResponseSelector(bow)`<br> test: `38s`, train: `4m34s`, total: `5m11s`|0.8960 (0.00)|0.9574 (0.00)|`no data`|
| `Sparse + Spacy + DIET(seq) + ResponseSelector(t2t)`<br> test: `57s`, train: `4m21s`, total: `5m18s`|0.8929 (-0.01)|0.9700 (0.00)|`no data`|


Dataset: `Private 2`, Dataset repository branch: `architecture-revamp/modified-spacy-configs`, commit: `a88501268464439d1d058a8b77f16a1c752230bd`
| Configuration | Intent Classification Micro F1 | Entity Recognition Micro F1 | Response Selection Micro F1 |
|---------------|-----------------|-----------------|-------------------|
| `BERT + DIET(bow) + ResponseSelector(bow)`<br> test: `1m50s`, train: `11m17s`, total: `13m7s`|0.8745 (0.00)|`no data`|`no data`|
| `BERT + DIET(seq) + ResponseSelector(t2t)`<br> test: `2m5s`, train: `6m18s`, total: `8m23s`|0.8830 (0.00)|`no data`|`no data`|
| `Spacy + DIET(bow) + ResponseSelector(bow)`<br> test: `38s`, train: `5m59s`, total: `6m36s`|0.7253 (-0.00)|`no data`|`no data`|
| `Spacy + DIET(seq) + ResponseSelector(t2t)`<br> test: `47s`, train: `6m0s`, total: `6m47s`|0.7822 (0.00)|`no data`|`no data`|
| `Sparse + DIET(bow) + ResponseSelector(bow)`<br> test: `37s`, train: `5m41s`, total: `6m18s`|0.8498 (0.01)|`no data`|`no data`|
| `Sparse + DIET(seq) + ResponseSelector(t2t)`<br> test: `44s`, train: `5m40s`, total: `6m24s`|0.8530 (-0.00)|`no data`|`no data`|
| `Sparse + Spacy + DIET(bow) + ResponseSelector(bow)`<br> test: `48s`, train: `8m35s`, total: `9m23s`|0.8594 (-0.00)|`no data`|`no data`|
| `Sparse + Spacy + DIET(seq) + ResponseSelector(t2t)`<br> test: `52s`, train: `7m16s`, total: `8m8s`|0.8509 (0.00)|`no data`|`no data`|


Dataset: `Private 3`, Dataset repository branch: `architecture-revamp/modified-spacy-configs`, commit: `a88501268464439d1d058a8b77f16a1c752230bd`
| Configuration | Intent Classification Micro F1 | Entity Recognition Micro F1 | Response Selection Micro F1 |
|---------------|-----------------|-----------------|-------------------|
| `BERT + DIET(bow) + ResponseSelector(bow)`<br> test: `59s`, train: `1m10s`, total: `2m8s`|0.9177 (0.00)|`no data`|`no data`|
| `BERT + DIET(seq) + ResponseSelector(t2t)`<br> test: `1m4s`, train: `52s`, total: `1m55s`|0.8436 (0.00)|`no data`|`no data`|
| `Spacy + DIET(bow) + ResponseSelector(bow)`<br> test: `36s`, train: `54s`, total: `1m30s`|0.6132 (-0.00)|`no data`|`no data`|
| `Spacy + DIET(seq) + ResponseSelector(t2t)`<br> test: `42s`, train: `44s`, total: `1m25s`|0.6255 (0.00)|`no data`|`no data`|
| `Sparse + DIET(bow) + ResponseSelector(bow)`<br> test: `33s`, train: `1m6s`, total: `1m38s`|0.8683 (0.00)|`no data`|`no data`|
| `Sparse + DIET(seq) + ResponseSelector(t2t)`<br> test: `37s`, train: `49s`, total: `1m25s`|0.8642 (0.00)|`no data`|`no data`|
| `Sparse + Spacy + DIET(bow) + ResponseSelector(bow)`<br> test: `37s`, train: `1m18s`, total: `1m55s`|0.8436 (-0.00)|`no data`|`no data`|
| `Sparse + Spacy + DIET(seq) + ResponseSelector(t2t)`<br> test: `43s`, train: `58s`, total: `1m40s`|0.8601 (0.00)|`no data`|`no data`|


Dataset: `Sara`, Dataset repository branch: `architecture-revamp/modified-spacy-configs`, commit: `a88501268464439d1d058a8b77f16a1c752230bd`

| Dialog Policy Configuration | Action Level Micro Avg. F1 | Conversation Level Accuracy | Run Time Train | Run Time Test |
|---------------|-----------------|-----------------|-------------------|-------------------|
| `Rules` |0.1266 (0.00)|0.0000 (0.00)|`2m34s`| `1m24s`|
| `Rules + AugMemo` |0.9149 (0.00)|0.6301 (0.00)|`2m30s`| `1m33s`|
| `Rules + AugMemo + TED` |0.9727 (-0.00)|0.7534 (-0.00)|`13m8s`| `3m6s`|
| `Rules + Memo` |0.3860 (-0.11)|0.1438 (-0.03)|`2m34s`| `1m27s`|
| `Rules + Memo + TED` |0.9555 (-0.00)|0.6644 (-0.00)|`13m2s`| `3m5s`|

**Pull request merged**